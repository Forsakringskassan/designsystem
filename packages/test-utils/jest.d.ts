export * from "./dist/types/jest";

declare global {
    namespace jest {
        interface Matchers<R> {
            /**
             * Expects element to have focus.
             */
            toHaveFocus(): R;
        }
    }
}
