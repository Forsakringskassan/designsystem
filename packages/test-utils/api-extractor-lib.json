{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
