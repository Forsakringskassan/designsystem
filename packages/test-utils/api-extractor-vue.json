{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
