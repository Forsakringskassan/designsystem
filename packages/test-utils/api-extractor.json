{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
