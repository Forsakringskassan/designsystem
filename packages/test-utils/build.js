const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
