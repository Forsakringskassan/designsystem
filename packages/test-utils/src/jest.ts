/* imported for side-effects only */
import "./matchers";

export * from "./lib";
