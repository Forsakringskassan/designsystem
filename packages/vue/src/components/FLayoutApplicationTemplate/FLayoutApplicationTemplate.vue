<script lang="ts">
import { defineComponent } from "vue";
import { isSet } from "@fkui/logic";

export default defineComponent({
    name: "FLayoutApplicationTemplate",
    computed: {
        showHeader(): boolean {
            return this.hasSlot("header");
        },
        showTopNavigation(): boolean {
            return this.hasSlot("top-navigation");
        },
        showFooter(): boolean {
            return this.hasSlot("footer");
        },
    },
    mounted() {
        document.body.classList.add("layout-application-template__body");
    },
    beforeUnmount() {
        document.body.classList.remove("layout-application-template__body");
    },
    methods: {
        hasSlot(name: string): boolean {
            return isSet(this.$slots[name]);
        },
    },
});
</script>

<template>
    <div class="layout-application-template">
        <header v-if="showHeader || showTopNavigation" ref="header" class="layout-application-template__header">
            <!--
@slot Slot for displaying the header.
        -->
            <slot v-if="showHeader" name="header"></slot>
            <nav v-if="showTopNavigation">
                <!--
@slot Slot for displaying top navigation.
            -->
                <slot name="top-navigation"></slot>
            </nav>
        </header>
        <main ref="primary-content" class="layout-application-template__main">
            <!--
@slot Slot for displaying the primary content.
            -->
            <slot name="default"></slot>
            <footer v-if="showFooter" class="layout-application-template__footer">
                <!--
@slot Slot for displaying the footer.
        -->
                <slot name="footer"></slot>
            </footer>
        </main>
    </div>
</template>
