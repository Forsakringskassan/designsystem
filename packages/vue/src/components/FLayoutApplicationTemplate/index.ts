export { default as FLayoutApplicationTemplate } from "./FLayoutApplicationTemplate.vue";
