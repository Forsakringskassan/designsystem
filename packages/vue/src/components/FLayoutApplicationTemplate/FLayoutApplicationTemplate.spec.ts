import "html-validate/jest";
import { mount, VueWrapper } from "@vue/test-utils";
import FLayoutApplicationTemplate from "./FLayoutApplicationTemplate.vue";

let wrapper: VueWrapper;

function createPlaceholderInDocument(): HTMLElement {
    const elem = document.createElement("div");
    if (document.body) {
        document.body.appendChild(elem);
    }
    return elem;
}

interface IApplicationTemplateSlots {
    header?: string;
    "top-navigation": string;
    default: string;
    footer?: string;
}

const defaultSlots: IApplicationTemplateSlots = {
    header: "HEADER",
    "top-navigation": "TOPNAVIGATION",
    default: "DEFAULT",
    footer: "FOOTER",
};

function createWrapper(slots: IApplicationTemplateSlots): VueWrapper {
    return mount(FLayoutApplicationTemplate, {
        slots: {
            ...slots,
        },
        attachTo: createPlaceholderInDocument(),
    });
}

describe("snapshots", () => {
    it("should match snapshot with all slots used", () => {
        wrapper = createWrapper(defaultSlots);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should render snapshot with no header and footer if running under another application", () => {
        const withoutOptsSlots: IApplicationTemplateSlots = {
            "top-navigation": "TOPNAVIGATION",
            default: "DEFAULT",
        };
        wrapper = createWrapper(withoutOptsSlots);
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("<header>", () => {
    it("should not insert <header> element if neither header or top-navigation slot is set", () => {
        expect.assertions(1);
        const wrapper = mount(FLayoutApplicationTemplate);
        const header = wrapper.find("header");
        expect(header.exists()).toBeFalsy();
    });

    it("should not insert <header> element if header or top-navigation slot is set but empty", () => {
        expect.assertions(1);
        const wrapper = mount(FLayoutApplicationTemplate, {
            header: "",
            "top-navigation": "",
        });
        const header = wrapper.find("header");
        expect(header.exists()).toBeFalsy();
    });

    it("should insert <header> element if header slot is present and non-empty", () => {
        expect.assertions(1);
        const wrapper = mount(FLayoutApplicationTemplate, {
            slots: {
                header: "lorem ipsum",
            },
        });
        const header = wrapper.find("header");
        expect(header.exists()).toBeTruthy();
    });

    it("should insert <header> element if top-navigation slot is present and non-empty", () => {
        expect.assertions(1);
        const wrapper = mount(FLayoutApplicationTemplate, {
            slots: {
                "top-navigation": "lorem ipsum",
            },
        });
        const header = wrapper.find("header");
        expect(header.exists()).toBeTruthy();
    });
});

it("should apply css class 'layout-application-template__body' to body-element", () => {
    wrapper = createWrapper(defaultSlots);
    expect(window.document.body.classList).toContain(
        "layout-application-template__body",
    );
});

describe("html-validate", () => {
    it("should allow defined slots", () => {
        expect.assertions(1);
        const slotTemplates = Object.entries(defaultSlots).map(
            ([key, value]) => `<template #${key}>${value}</template>`,
        );

        expect(
            `<f-layout-application-template>${slotTemplates}</f-layout-application-template>`,
        ).toHTMLValidate();
    });
});
