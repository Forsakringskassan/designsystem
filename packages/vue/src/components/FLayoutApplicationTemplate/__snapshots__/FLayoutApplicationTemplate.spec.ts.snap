// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with all slots used 1`] = `
<div
  class="layout-application-template"
>
  <header
    class="layout-application-template__header"
  >
    
    HEADER
    
    <nav>
      
      TOPNAVIGATION
      
    </nav>
  </header>
  <main
    class="layout-application-template__main"
  >
    
    DEFAULT
    
    <footer
      class="layout-application-template__footer"
    >
      
      FOOTER
      
    </footer>
  </main>
</div>
`;

exports[`snapshots should render snapshot with no header and footer if running under another application 1`] = `
<div
  class="layout-application-template"
>
  <header
    class="layout-application-template__header"
  >
    <!--v-if-->
    <nav>
      
      TOPNAVIGATION
      
    </nav>
  </header>
  <main
    class="layout-application-template__main"
  >
    
    DEFAULT
    
    <!--v-if-->
  </main>
</div>
`;
